src/lib.rs

```diff
@@ -0,0 +1,371 @@
+use lazy_static::lazy_static;
+use regex::Regex;
+use resmoke_proxy::ResmokeProxy;
+use shrub_rs::models::{
+    project::EvgProject,
+    task::{EvgTask, TaskRef},
+    variant::{BuildVariant, DisplayTask},
+};
+use task_name::remove_gen_suffix;
+use task_types::{
+    fuzzer_tasks::{FuzzerGenTaskParams, GenFuzzerService, GenFuzzerServiceImpl},
+    generated_suite::GeneratedSuite,
+    multiversion::MultiversionServiceImpl,
+};
+use tracing::{event, Level};
+
+mod evergreen_names;
+mod evg_config;
+mod resmoke_proxy;
+mod task_name;
+mod task_types;
+
+/// Directory to store the generated configuration in.
+const CONFIG_DIR: &str = "generated_resmoke_config";
+
+lazy_static! {
+    static ref EXPANSION_RE: Regex =
+        Regex::new(r"\$\{(?P<id>[a-zA-Z0-9_]+)(\|(?P<default>.*))?}").unwrap();
+}
+
+type GenTaskCollection = HashMap<String, Box<dyn GeneratedSuite>>;
+
+/// Collection of services needed to execution.
+#[derive(Clone)]
+pub struct Dependencies {
+    gen_task_service: Arc<dyn GenerateTasksService>,
+}
+
+impl Dependencies {
+    /// Create a new set of dependency instances.
+    ///
+    /// # Arguments
+    ///
+    /// * `evg_project_location` - Path to the evergreen project configuration yaml.
+    ///
+    /// # Returns
+    ///
+    /// A set of dependencies to run against.
+    pub fn new(evg_project_location: &Path) -> Result<Self> {
+        let discovery_service = Arc::new(ResmokeProxy::new());
+        let multiversion_service = Arc::new(MultiversionServiceImpl::new(discovery_service)?);
+        let evg_config_service = Arc::new(EvgProjectConfig::new(evg_project_location)?);
+        let evg_config_utils = Arc::new(EvgConfigUtilsImpl::new());
+        let gen_fuzzer_service = Arc::new(GenFuzzerServiceImpl::new(multiversion_service));
+        let gen_task_service = Arc::new(GenerateTasksServiceImpl::new(
+            evg_config_service,
+            evg_config_utils,
+            gen_fuzzer_service,
+        ));
+
+        Ok(Self { gen_task_service })
+    }
+}
+
+/// A container for configuration generated for a build variant.
+#[derive(Debug, Clone)]
+struct GeneratedConfig {
+    /// References to generated tasks that should be included.
+    pub gen_task_specs: Vec<TaskRef>,
+    /// Display tasks that should be created.
+    pub display_tasks: Vec<DisplayTask>,
+}
+
+impl GeneratedConfig {
+    /// Create an empty instance of generated configuration.
+    pub fn new() -> Self {
+        Self {
+            gen_task_specs: vec![],
+            display_tasks: vec![],
+        }
+    }
+}
+
+/// Create 'generate.tasks' configuration for all generated tasks in the provided evergreen
+/// project configuration.
+///
+/// # Arguments
+///
+/// * `deps` - Dependencies needed to perform generation.
+/// * `config_location` - Pointer to S3 location that configuration will be uploaded to.
+pub fn generate_configuration(deps: Dependencies, config_location: &str) -> Result<()> {
```

Comment:
`()` is a unit type. Here it means this function doesn't return anything. In this case, the function won't return anything, but it could fail. So we want to return a `Result` in case there is an error. The `()` is used for that. If there is an error the `Result` will be an error, if everything was successful, the result will be `()`.